### PR TITLE
Multiply Decorated Cospans

### PR DESCRIPTION
**Pull request by mehalter**
Added multiply decorated morphisms and made RelOlogModels and PetriModels AbstractModels

### CODE REVIEW

**Comment by jpfairbanks** [comment]
I just tried to run the demo and got this error 
```
MethodError: no method matching ⊔(::Array{Any,1}, ::Array{Any,1})

Stacktrace:
 [1] pushout(::Span{Decorated{FinSetMorph{UnitRange{Int64},Array{Int64,1}},Array{Any,1}},Decorated{FinSetMorph{UnitRange{Int64},Array{Int64,1}},Array{Any,1}}}) at /Users/jfairbanks6/.julia/dev/SemanticModels/src/modeltools/CategoryTheory.jl:294
 [2] top-level scope at In[43]:3
```

**Comment by jpfairbanks** [comment]
I think we might want to consider thinking of a MultiplyDecorated Cosan equivalent to a  Singly Decorated Cospan where the decoration happens to be a tuple.

**Comment by mehalter** [comment]
I think how it is now represents that well. A singly decorated cospan is currently a multiplydecorated cospan with a single decoration

**Comment by mehalter** [comment]
What error are you talking about? I cannot recreate. I restart the kernel and rebuild and it works.

**Comment by jpfairbanks** [comment]
Restarting the kernel fixed it

File: src/modeltools/CategoryTheory.jl [hunk 1]
```diff
@@ -1,24 +1,21 @@
...
+export ⊔, AbstractMorph, FinSetMorph, dom, codom, verify, func, Decorated, decorations, undecorate, add_decoration!, add_decorations!, remove_decoration!, remove_decorations!, AbstractSpan, leftob, rightob, apexob, Span, left, right, DoublePushout, AbstractCospan, Cospan, pushout
 
-# +
 import MacroTools.walk
 walk(x::Operation, inner, outer) = outer(Operation(x.op, map(inner, x.args)))
 
 
 ⊔(a::UnitRange, b::UnitRange) = 1:(length(a)+length(b))
 ⊔(a::AbstractVector{Int}, b::AbstractVector{Int}) = vcat(a,b)
-⊔(g::AbstractGraph, h::AbstractGraph) = blockdiag(g,h)
```
Comment (jpfairbanks) [inline src/modeltools/CategoryTheory.jl]: Did this go somewhere else?

Comment (jpfairbanks) [reply src/modeltools/CategoryTheory.jl]: Found it

File: src/modeltools/CategoryTheory.jl [hunk 2]
```diff
@@ -71,81 +67,58 @@ function ⊔(f::FinSetMorph, g::FinSetMorph)
...
-    end |> all
+# Add a decoration to a decorated morphism
+function add_decoration!(m::Decorated, decoration::AbstractModel)
+  push!(decorations(m), decoration)
 end
 
-# +
-"""    Decorated{M,T}
+# Add a collection of decorations to a decorated morphism
+function add_decorations!(m::Decorated, decorations)
```
Comment (jpfairbanks) [inline src/modeltools/CategoryTheory.jl]: I think this should be `Base.append!(m::Decorated, decorations) = append!(m.d, decorations)` for consistency with the standard library naming convention?

File: src/modeltools/CategoryTheory.jl [hunk 3]
```diff
@@ -71,81 +67,58 @@ function ⊔(f::FinSetMorph, g::FinSetMorph)
...
-    f = func(m)
-    map(edges(dom(m))) do e
-        u,v = f(e.src), f(e.dst)
-        if u > v
-            u,v = v,u
-        end
-        Edge(u,v) in E
-    end |> all
+# Add a decoration to a decorated morphism
+function add_decoration!(m::Decorated, decoration::AbstractModel)
```
Comment (jpfairbanks) [inline src/modeltools/CategoryTheory.jl]: We should use multiple dispatch here to make this `Base.push!(m::Decorated, decoration::AbstractModel) = push!(decorations(m), decoration)`

File: src/modeltools/CategoryTheory.jl [hunk 4]
```diff
@@ -71,81 +67,58 @@ function ⊔(f::FinSetMorph, g::FinSetMorph)
...
+end
 
-a decoration applied to the objects of a morphism, where M is a type of morphism and
-type T is the category of the decoration
-"""
-struct Decorated{M,T}
-    f::M
-    d::T
+# remove a decoration from a decorated morphism
+function remove_decoration!(m::Decorated, i)
```
Comment (jpfairbanks) [inline src/modeltools/CategoryTheory.jl]: `Base.deleteat!(m::Decorated, i) = deleteat!(decorations(m),i)`

File: src/modeltools/CategoryTheory.jl (hunk 2 above)
Comment (mehalter) [reply src/modeltools/CategoryTheory.jl]: done.

File: src/modeltools/CategoryTheory.jl (hunk 3 above)
Comment (mehalter) [reply src/modeltools/CategoryTheory.jl]: done.

File: src/modeltools/CategoryTheory.jl (hunk 4 above)
Comment (mehalter) [reply src/modeltools/CategoryTheory.jl]: done.

**Pull request merged**